stan/math/prim/fun/LDLT_factor.hpp

```diff
@@ -12,120 +12,69 @@ namespace stan {
-    compute(A);
-  }
-
-  inline void compute(const matrix_t& A) {
-    check_square("LDLT_factor", "A", A);
-    N_ = A.rows();
-    ldltP_->compute(A);
-  }
-
-  inline bool success() const {
-    if (ldltP_->info() != Eigen::Success) {
-      return false;
-    }
-    if (!(ldltP_->isPositive())) {
-      return false;
-    }
-    vector_t ldltP_diag(ldltP_->vectorD());
-    for (int i = 0; i < ldltP_diag.size(); ++i) {
-      if (ldltP_diag(i) <= 0 || is_nan(ldltP_diag(i))) {
-        return false;
-      }
-    }
-    return true;
-  }
-
-  inline T log_abs_det() const { return sum(log(ldltP_->vectorD())); }
+template <typename T, bool alloc_in_arena>
+class LDLT_factor;
 
-  inline void inverse(matrix_t& invA) const {
-    invA.setIdentity(N_);
-    ldltP_->solveInPlace(invA);
-  }
-
-  template <typename Rhs>
-  inline const Eigen::Solve<ldlt_t, Rhs> solve(
-      const Eigen::MatrixBase<Rhs>& b) const {
-    return ldltP_->solve(b);
-  }
-
-  inline matrix_t solveRight(const matrix_t& B) const {
-    return ldltP_->solve(B.transpose()).transpose();
-  }
-
-  inline vector_t vectorD() const { return ldltP_->vectorD(); }
-
-  inline ldlt_t matrixLDLT() const { return ldltP_->matrixLDLT(); }
-
-  inline size_t rows() const { return N_; }
-  inline size_t cols() const { return N_; }
+/**
+ * An LDLT_factor with `alloc_in_arena = False` is a structure that
+ * holds a matrix of type T and the LDLT of its values.
+ *
+ * @tparam T type of elements in the matrix
+ */
+template <typename T>
+class LDLT_factor<Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>, false> {
+ private:
+  using ldlt_type
+      = Eigen::LDLT<Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>>;
+  std::shared_ptr<ldlt_type> ldlt_ptr_;
 
-  size_t N_;
-  std::shared_ptr<ldlt_t> ldltP_;
+ public:
+  template <typename S, require_eigen_t<S>* = nullptr>
+  explicit LDLT_factor(const S& matrix) : ldlt_ptr_(new ldlt_type(matrix)) {}
+
+  /**
+   * Return a const reference to the underlying matrix
+   */
+  const auto& matrix() const { return ldlt_ptr_->matrixLDLT(); }
+
+  /**
+   * Return a const reference to the LDLT factor of the matrix
+   */
+  const auto& ldlt() const { return *ldlt_ptr_; }
 };
 
+/**
+ * Make an LDLT_factor with matrix type `T` and allocate any necessary
+ * memory in the arena if any of the types `T` or `Args...` contain
+ * `var_value<T>` types.
+ *
+ * @tparam T Type of matrix to take the LDLT of
+ * @tparam Args Other arguments to decide where to allocate LDLT
+ * @param A Matrix to take the LDLT of
+ * @return LDLT_factor of A
+ */
+template <typename T, typename... Args, require_matrix_t<T>* = nullptr>
+inline auto make_ldlt_factor(const T& A) {
+  return LDLT_factor<
+      plain_type_t<T>,
+      disjunction<is_var<scalar_type_t<return_type_t<T, Args...>>>,
+                  is_var<partials_type_t<return_type_t<Args...>>>,
+                  is_var<partials_type_t<
+                      partials_type_t<return_type_t<Args...>>>>>::value>(A);
```

Comment:
I'm guessing this is for `var`, `fvar<var>` and `fvar<fvar<var>>`. Maybe we should just have a type trait like `contains_var` that recurses through `fvar<>`?

Reply:
Yeah, this is what it is doing, and yes, that sounds handy for this, though I'm not sure we'd use it anywhere other than here.

Normally in Stan an operation is responsible for saving everything it needs in the reverse pass to the arena. `LDLT_factor` allocates itself in a way that it will be saved if it needs saved (instead of leaving this up to the op). I kinda want to change it, though it might make the functions a bit slower (cuz there'd be extra copies).

Reply:
Reading this again I want to try removing `alloc_in_arena` and see how expensive that actually ends up being. It's kindof an annoying thing.

Reply:
You could just have it be an internal type trait only defined here in ldlt_factor (lest we think it would be used elsewhere)

Reply:
I got rid of this type checking.